[6.2] use DatabaseFactory in DatabaseServiceProvider

heelc29 opened the pull request:
### Summary of Changes

`DatabaseDriver::getInstance` is deprecated and should be replaced with `DatabaseFactory->getDriver()`
https://github.com/joomla-framework/database/blob/d13a7504471b4730a8107e3566c6fe6d476706fc/src/DatabaseDriver.php#L269-L288

### Testing Instructions

install Joomla and visit frontend/backend and see if you can do anything (connection to database is present)

### Actual result BEFORE applying this Pull Request

Warning in deprecated log: `DatabaseDriver::getInstance` is deprecated and will be removed in 3.0, use `DatabaseFactory::getDriver()` instead.

```
 ------ -------------------------------------------------- 
  Line   libraries/src/Service/Provider/Database.php       
 ------ -------------------------------------------------- 
  123    Call to deprecated method getInstance() of class  
         Joomla\Database\DatabaseDriver:                   
         3.0  Use DatabaseFactory::getDriver() instead     
         🪪  staticMethod.deprecated                       
 ------ -------------------------------------------------- 
```

### Expected result AFTER applying this Pull Request

Warning gone

### Link to documentations
Please select:
- [x] No documentation changes for docs.joomla.org needed
- [x] No documentation changes for manual.joomla.org needed

heelc29 committed:
[5.3] use DatabaseFactory in DatabaseServiceProvider

heelc29 commented:
Failing CI related to #45059
```
 Error: Unwanted PHP Deprecated: "  htmlspecialchars(): Passing null to parameter #1 ($string) of type string is deprecated in <b>/tests/www/cmysql/administrator/modules/mod_logged/tmpl/default.php</b> on line <b>36</b>"
```

voronkovich commented on libraries/src/Service/Provider/Database.php:
```diff
@@ -120,7 +121,7 @@ function (Container $container) {
 
                     try {
-                        $db = DatabaseDriver::getInstance($options);
+                        $db = $options['factory']->getDriver($dbtype, $options);
```
The `factory` option is set automatically by `DatabaseFactory`. I would suggest not changing the `$options` array and instantiating the database like this:

```php
$db = (new DatabaseFactory())->getDriver($dbtype, $options);
```

heelc29 replied:
Done. Thanks.

heelc29 committed:
remove factory from option array

Co-Authored-By: Oleg Voronkovich <oleg-voronkovich@yandex.ru>

exlemor commented:
Hi @heelc29, I might be doing something incorrectly or the testing instructions are incomplete or I couldn't replicate the BEFORE condition as I could not find in the Deprecated.php log any mention of Call to deprecated method getInstance() of class  
         Joomla\Database\DatabaseDriver:                   
         3.0  Use DatabaseFactory::getDriver() instead     
         🪪  staticMethod.deprecated    
         
You night want to update your Testing Instructions to advise the turn: Log Deprecated API in Global Configuration if that is where a tester is supposed to find this information...

heelc29 committed:
use DatabaseFactory

heelc29 commented:
Update more `DatabaseDriver::getInstance`

Also add system test for config cli command

heelc29 committed:
Merge branch '5.3-dev' into 5.3/database/provider

heelc29 committed:
fix

heelc29 committed:
Merge branch '5.3-dev' into 5.3/database/provider

heelc29 committed:
Merge branch '5.3-dev' into 5.3/database/provider

heelc29 committed:
fix

heelc29 committed:
Merge branch '5.3-dev' into 5.3/database/provider

heelc29 committed:
Merge branch '6.1-dev' into 5.3/database/provider

laoneo committed:
Merge branch '6.1-dev' into 5.3/database/provider

heelc29 committed:
Merge branch '6.1-dev' into 5.3/database/provider

heelc29 committed:
ignore formatting

heelc29 committed:
cy.exec code property renamed

https://docs.cypress.io/app/references/migration-guide#cyexec-code-property-renamed

heelc29 commented:
> @heelc29 the system test error seems to be related.

@HLeithner fixed

laoneo committed:
Merge branch '6.2-dev' into 5.3/database/provider

heelc29 committed:
update test

see #46876

heelc29 committed:
Update phpstan-baseline.neon

heelc29 committed:
Merge branch '6.2-dev' into 5.3/database/provider

heelc29 committed:
Merge branch '6.2-dev' into 5.3/database/provider

laoneo committed:
Merge branch '6.2-dev' into 5.3/database/provider
